In [1]:
#남산도서관 데이터 다운
import gdown
gdown.download('https://bit.ly/3RhoNho', 'ns_202104.csv',quiet=False)

Downloading...
From: https://bit.ly/3RhoNho
To: /content/ns_202104.csv
100%|██████████| 57.6M/57.6M [00:00<00:00, 101MB/s] 


'ns_202104.csv'

In [2]:
#판다스 데이터프레임으로 첫 다섯 행 출력
import pandas as pd
ns_df = pd.read_csv('ns_202104.csv', low_memory=False)
ns_df.head()

번호                    도서명                저자    출판사  발행년도           ISBN  \
0   1                인공지능과 흙            김동훈 지음    민음사  2021  9788937444319   
1   2           가짜 행복 권하는 사회            김태형 지음   갈매나무  2021  9791190123969   
2   3  나도 한 문장 잘 쓰면 바랄 게 없겠네            김선영 지음   블랙피쉬  2021  9788968332982   
3   4                예루살렘 해변  이도 게펜 지음, 임재희 옮김  문학세계사  2021  9788970759906   
4   5  김성곤의 중국한시기행 : 장강·황하 편            김성곤 지음    김영사  2021  9788934990833   

  세트 ISBN 부가기호    권 주제분류번호  도서권수  대출건수        등록일자  Unnamed: 13  
0     NaN  NaN  NaN    NaN     1     0  2021-03-19          NaN  
1     NaN  NaN  NaN    NaN     1     0  2021-03-19          NaN  
2     NaN  NaN  NaN    NaN     1     0  2021-03-19          NaN  
3     NaN  NaN  NaN    NaN     1     0  2021-03-19          NaN  
4     NaN  NaN  NaN    NaN     1     0  2021-03-19          NaN

### 열 삭제하기
1. loc메서드로 필요한 열 선택 with 슬라이스 연산자/불리언 배열)<br>
2. drop메서드 with axis=1 <br>
삭제할 열 여러개면 열 이름 리스트로 전달 가능(이게 여기 설명 맞나? 확인)
3. dropna()메서드 axis=1하면 NaN이 들어있는 열 삭제 가능

In [3]:
#판다스가 자동으로 추가한 불필요한 Unnamed13열 삭제하기
#loc메서드+슬라이싱 사용하여 '번호'~'등록일자'까지 선택해서 새로운 데이터프레임 만들기
ns_book = ns_df.loc[:, '번호':'등록일자']
ns_book.head()

번호                    도서명                저자    출판사  발행년도           ISBN  \
0   1                인공지능과 흙            김동훈 지음    민음사  2021  9788937444319   
1   2           가짜 행복 권하는 사회            김태형 지음   갈매나무  2021  9791190123969   
2   3  나도 한 문장 잘 쓰면 바랄 게 없겠네            김선영 지음   블랙피쉬  2021  9788968332982   
3   4                예루살렘 해변  이도 게펜 지음, 임재희 옮김  문학세계사  2021  9788970759906   
4   5  김성곤의 중국한시기행 : 장강·황하 편            김성곤 지음    김영사  2021  9788934990833   

  세트 ISBN 부가기호    권 주제분류번호  도서권수  대출건수        등록일자  
0     NaN  NaN  NaN    NaN     1     0  2021-03-19  
1     NaN  NaN  NaN    NaN     1     0  2021-03-19  
2     NaN  NaN  NaN    NaN     1     0  2021-03-19  
3     NaN  NaN  NaN    NaN     1     0  2021-03-19  
4     NaN  NaN  NaN    NaN     1     0  2021-03-19

**불리언 배열** : 열을 불리언 배열로 나타내는 것<br>
중간에 있는 '부가기호'열만 제외시키기

In [4]:
#먼저 데이터프레임 열 이름이 저장된 columns 속성 확인
print(ns_df.columns)

Index(['번호', '도서명', '저자', '출판사', '발행년도', 'ISBN', '세트 ISBN', '부가기호', '권',
       '주제분류번호', '도서권수', '대출건수', '등록일자', 'Unnamed: 13'],
      dtype='object')


columns속성=판다스의 index 객체 -> 리스트처럼 **숫자인덱스** 로 참조 가능
```
print(ns_df.columns[0])
>>> 번호
```

# 원소별 비교
Index 클래스를 비롯한 **판다스 배열 성격의 객체** <br>
= 어떤 값과 비교할 때 **자동으로 배열에 있는 모든 원소와 하나씩 비교**해 줌<br>
ex. 인덱스 ["혼공", "분석", "파이썬"]과 문자열 "분석"을 비교하면 결과로 False, True, False 나열하여 반환


> 반환된 결과 = **넘파이 배열** <br> 파이썬 리스트와 비슷한 걸로 생각해도 됨




In [5]:
#Unnamed: 13'열이 아닌 것을 표시하는 배열
ns_df.columns != 'Unnamed: 13'
#반환된 결과가 넘파이 배열

array([ True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True, False])

In [8]:
#앞에서 구한 배열을 selected_columns 변수에 저장 -> 판다스 데이터프레임 loc메서드 전달->True인 행만 선택 가능
selected_columns = ns_df.columns != 'Unnamed: 13'
ns_book = ns_df.loc[:, selected_columns] #True인 모든 행 선택
#하면 unnamed:13 열 없어짐
ns_book.head()
#하면 앞에 슬라이싱이랑 똑같은 결과 나옴

번호                    도서명                저자    출판사  발행년도           ISBN  \
0   1                인공지능과 흙            김동훈 지음    민음사  2021  9788937444319   
1   2           가짜 행복 권하는 사회            김태형 지음   갈매나무  2021  9791190123969   
2   3  나도 한 문장 잘 쓰면 바랄 게 없겠네            김선영 지음   블랙피쉬  2021  9788968332982   
3   4                예루살렘 해변  이도 게펜 지음, 임재희 옮김  문학세계사  2021  9788970759906   
4   5  김성곤의 중국한시기행 : 장강·황하 편            김성곤 지음    김영사  2021  9788934990833   

  세트 ISBN 부가기호    권 주제분류번호  도서권수  대출건수        등록일자  
0     NaN  NaN  NaN    NaN     1     0  2021-03-19  
1     NaN  NaN  NaN    NaN     1     0  2021-03-19  
2     NaN  NaN  NaN    NaN     1     0  2021-03-19  
3     NaN  NaN  NaN    NaN     1     0  2021-03-19  
4     NaN  NaN  NaN    NaN     1     0  2021-03-19

In [9]:
#부가기호 열 제거
selected_columns = ns_df.columns != '부가기호'
ns_book = ns_df.loc[:, selected_columns]
ns_book.head()

번호                    도서명                저자    출판사  발행년도           ISBN  \
0   1                인공지능과 흙            김동훈 지음    민음사  2021  9788937444319   
1   2           가짜 행복 권하는 사회            김태형 지음   갈매나무  2021  9791190123969   
2   3  나도 한 문장 잘 쓰면 바랄 게 없겠네            김선영 지음   블랙피쉬  2021  9788968332982   
3   4                예루살렘 해변  이도 게펜 지음, 임재희 옮김  문학세계사  2021  9788970759906   
4   5  김성곤의 중국한시기행 : 장강·황하 편            김성곤 지음    김영사  2021  9788934990833   

  세트 ISBN    권 주제분류번호  도서권수  대출건수        등록일자  Unnamed: 13  
0     NaN  NaN    NaN     1     0  2021-03-19          NaN  
1     NaN  NaN    NaN     1     0  2021-03-19          NaN  
2     NaN  NaN    NaN     1     0  2021-03-19          NaN  
3     NaN  NaN    NaN     1     0  2021-03-19          NaN  
4     NaN  NaN    NaN     1     0  2021-03-19          NaN

**drop()메서드를 통해 열 삭제하기**<br>
데이터프레임의 행이나 열을 삭제하는 메서드<br>
첫 번째 메서드: 삭제하려는 열 이름 <br>
두 번째 메서드: axis매개변수 = 1


In [10]:
#Unnamed: 13열 삭제하기
ns_book = ns_df.drop('Unnamed: 13', axis=1)
ns_book.head()

번호                    도서명                저자    출판사  발행년도           ISBN  \
0   1                인공지능과 흙            김동훈 지음    민음사  2021  9788937444319   
1   2           가짜 행복 권하는 사회            김태형 지음   갈매나무  2021  9791190123969   
2   3  나도 한 문장 잘 쓰면 바랄 게 없겠네            김선영 지음   블랙피쉬  2021  9788968332982   
3   4                예루살렘 해변  이도 게펜 지음, 임재희 옮김  문학세계사  2021  9788970759906   
4   5  김성곤의 중국한시기행 : 장강·황하 편            김성곤 지음    김영사  2021  9788934990833   

  세트 ISBN 부가기호    권 주제분류번호  도서권수  대출건수        등록일자  
0     NaN  NaN  NaN    NaN     1     0  2021-03-19  
1     NaN  NaN  NaN    NaN     1     0  2021-03-19  
2     NaN  NaN  NaN    NaN     1     0  2021-03-19  
3     NaN  NaN  NaN    NaN     1     0  2021-03-19  
4     NaN  NaN  NaN    NaN     1     0  2021-03-19

In [11]:
#중간에 있는 '부가기호' 열과 끝의 'Unnamed: 13' 열 동시에 제거하기
ns_book = ns_df.drop(['부가기호','Unnamed: 13'],axis=1)
ns_book.head()

번호                    도서명                저자    출판사  발행년도           ISBN  \
0   1                인공지능과 흙            김동훈 지음    민음사  2021  9788937444319   
1   2           가짜 행복 권하는 사회            김태형 지음   갈매나무  2021  9791190123969   
2   3  나도 한 문장 잘 쓰면 바랄 게 없겠네            김선영 지음   블랙피쉬  2021  9788968332982   
3   4                예루살렘 해변  이도 게펜 지음, 임재희 옮김  문학세계사  2021  9788970759906   
4   5  김성곤의 중국한시기행 : 장강·황하 편            김성곤 지음    김영사  2021  9788934990833   

  세트 ISBN    권 주제분류번호  도서권수  대출건수        등록일자  
0     NaN  NaN    NaN     1     0  2021-03-19  
1     NaN  NaN    NaN     1     0  2021-03-19  
2     NaN  NaN    NaN     1     0  2021-03-19  
3     NaN  NaN    NaN     1     0  2021-03-19  
4     NaN  NaN    NaN     1     0  2021-03-19

그러니까, 중간에 있는 열을 빼고 나머지 열을 선택할 때는 불리언 배열이나 drop()메서드를 사용하면 됨

**drop() 메서드에 inplace 매개변수=True로 하면 선택한 데이터프레임 바로 수정 가능**<br>
새로운 변수에 데이터프레임 수정 결과를 반환하지 않고,<br>ns_book에서 바로 '주제분류번호 열을 삭제'하는 것

In [12]:
ns_book.drop('주제분류번호', axis=1, inplace=True)
ns_book.head()

번호                    도서명                저자    출판사  발행년도           ISBN  \
0   1                인공지능과 흙            김동훈 지음    민음사  2021  9788937444319   
1   2           가짜 행복 권하는 사회            김태형 지음   갈매나무  2021  9791190123969   
2   3  나도 한 문장 잘 쓰면 바랄 게 없겠네            김선영 지음   블랙피쉬  2021  9788968332982   
3   4                예루살렘 해변  이도 게펜 지음, 임재희 옮김  문학세계사  2021  9788970759906   
4   5  김성곤의 중국한시기행 : 장강·황하 편            김성곤 지음    김영사  2021  9788934990833   

  세트 ISBN    권  도서권수  대출건수        등록일자  
0     NaN  NaN     1     0  2021-03-19  
1     NaN  NaN     1     0  2021-03-19  
2     NaN  NaN     1     0  2021-03-19  
3     NaN  NaN     1     0  2021-03-19  
4     NaN  NaN     1     0  2021-03-19

**dropna()메서드**<br>
NaN:판다스가 비어있는 칸을 표현하는것<br>
dropna()메서드: NaN이 하나 이상 포함된 행이나 열 삭제


In [13]:
#axis=1로 지정하여 ns_df 데이터프레임에서 NaN이 포함된 열 삭제
ns_book = ns_df.dropna(axis=1)
ns_book.head()
#도서명, 저자 이런 곳에 한 곳에라오 NaN가 있었으면 다 삭제된거

번호           ISBN  도서권수  대출건수        등록일자
0   1  9788937444319     1     0  2021-03-19
1   2  9791190123969     1     0  2021-03-19
2   3  9788968332982     1     0  2021-03-19
3   4  9788970759906     1     0  2021-03-19
4   5  9788934990833     1     0  2021-03-19

In [14]:
#모든 값이 NaN인 열만 삭제하려면 dropna()메서드에 how매개변수='all'로 지정하면 됨
ns_book = ns_df.dropna(axis=1,how='all')
ns_book.head()

번호                    도서명                저자    출판사  발행년도           ISBN  \
0   1                인공지능과 흙            김동훈 지음    민음사  2021  9788937444319   
1   2           가짜 행복 권하는 사회            김태형 지음   갈매나무  2021  9791190123969   
2   3  나도 한 문장 잘 쓰면 바랄 게 없겠네            김선영 지음   블랙피쉬  2021  9788968332982   
3   4                예루살렘 해변  이도 게펜 지음, 임재희 옮김  문학세계사  2021  9788970759906   
4   5  김성곤의 중국한시기행 : 장강·황하 편            김성곤 지음    김영사  2021  9788934990833   

  세트 ISBN 부가기호    권 주제분류번호  도서권수  대출건수        등록일자  
0     NaN  NaN  NaN    NaN     1     0  2021-03-19  
1     NaN  NaN  NaN    NaN     1     0  2021-03-19  
2     NaN  NaN  NaN    NaN     1     0  2021-03-19  
3     NaN  NaN  NaN    NaN     1     0  2021-03-19  
4     NaN  NaN  NaN    NaN     1     0  2021-03-19

# 행 삭제하기

drop()메서드 사용<br>
axis=0하면 되지만, 0이 기본값이라 안써도 됨<br>
**열 삭제와 마찬가지로, loc 메서드, drop()메서드, dropna()메서드 모두 사용 가능<br> 다만, 불리언 배열을 [ ]안에 전달하는 방법을 많이 사용한다!**



In [15]:
#첫 두 행 삭제하기
ns_book2 = ns_book.drop([0,1]) #인덱스 0부터 1까지
ns_book2.head()

번호                                  도서명                    저자  \
2   3                나도 한 문장 잘 쓰면 바랄 게 없겠네                김선영 지음   
3   4                              예루살렘 해변      이도 게펜 지음, 임재희 옮김   
4   5                김성곤의 중국한시기행 : 장강·황하 편                김성곤 지음   
5   6                        처음 읽는 음식의 세계사  미야자키 마사카츠 지음, 한세희 옮김   
6   7  아르센 벵거 자서전 My Life in Red and White     아르센 벵거 지음, 이성모 옮김   

            출판사  발행년도           ISBN 세트 ISBN 부가기호    권 주제분류번호  도서권수  대출건수  \
2          블랙피쉬  2021  9788968332982     NaN  NaN  NaN    NaN     1     0   
3         문학세계사  2021  9788970759906     NaN  NaN  NaN    NaN     1     0   
4           김영사  2021  9788934990833     NaN  NaN  NaN    NaN     1     0   
5          탐나는책  2021  9791189550370     NaN  NaN  NaN    NaN     1     0   
6  한즈미디어(한스미디어)  2021  9791160075793     NaN  NaN  NaN    NaN     1     0   

         등록일자  
2  2021-03-19  
3  2021-03-19  
4  2021-03-19  
5  2021-03-19  
6  2021-03-19

In [16]:
#근데 숫자 인덱스로 지정하는건 처음 한두개만 하지 쉽지 않음
#[]에 슬라이싱이나 불리언 배열 넣으면 '행'(열아님)을 선택!

#인덱스 0,1인 행 제외한 모든 행 선택하기
#슬라이싱 사용
ns_book2 = ns_book[2:]
ns_book2.head()

번호                                  도서명                    저자  \
2   3                나도 한 문장 잘 쓰면 바랄 게 없겠네                김선영 지음   
3   4                              예루살렘 해변      이도 게펜 지음, 임재희 옮김   
4   5                김성곤의 중국한시기행 : 장강·황하 편                김성곤 지음   
5   6                        처음 읽는 음식의 세계사  미야자키 마사카츠 지음, 한세희 옮김   
6   7  아르센 벵거 자서전 My Life in Red and White     아르센 벵거 지음, 이성모 옮김   

            출판사  발행년도           ISBN 세트 ISBN 부가기호    권 주제분류번호  도서권수  대출건수  \
2          블랙피쉬  2021  9788968332982     NaN  NaN  NaN    NaN     1     0   
3         문학세계사  2021  9788970759906     NaN  NaN  NaN    NaN     1     0   
4           김영사  2021  9788934990833     NaN  NaN  NaN    NaN     1     0   
5          탐나는책  2021  9791189550370     NaN  NaN  NaN    NaN     1     0   
6  한즈미디어(한스미디어)  2021  9791160075793     NaN  NaN  NaN    NaN     1     0   

         등록일자  
2  2021-03-19  
3  2021-03-19  
4  2021-03-19  
5  2021-03-19  
6  2021-03-19

In [17]:
#[]에 슬라이싱 쓰면 loc이랑 다르게 마지막 인덱스 포함하지 않음
ns_book2 = ns_book[0:2]
ns_book2.head()
#처음 두 개 행만 나옴

번호           도서명      저자   출판사  발행년도           ISBN 세트 ISBN 부가기호    권  \
0   1       인공지능과 흙  김동훈 지음   민음사  2021  9788937444319     NaN  NaN  NaN   
1   2  가짜 행복 권하는 사회  김태형 지음  갈매나무  2021  9791190123969     NaN  NaN  NaN   

  주제분류번호  도서권수  대출건수        등록일자  
0    NaN     1     0  2021-03-19  
1    NaN     1     0  2021-03-19

**불리언 배열 사용한 행 선택**<br>
행 선택할 때 가장 많이 쓰이는 방법<br>
원하는 행=True / 제외할 행=False

In [18]:
#출판사='한빛미디어'인 행만 선택
#순서1)비교연산자 사용하여 원본 데이터프레임 ns_df에서 출판사 '한빛미디어'인 행만 True인 불리언 배열 만들기
#순서2)불리언 배열을 []연산자에 넣기
selected_rows = ns_df['출판사']=='한빛미디어'
ns_book2 = ns_book[selected_rows]
ns_book2.head()

번호                                                도서명  \
60    61       (맛있는 디자인)프리미어 프로 CC: 쉽게 배워 제대로 써먹는 유튜브 영상 편집   
70    71         처음 배우는 애저 (Azure Portal로 배우는 애저 도입부터 활용까지)   
88    89  맛있는 디자인 프리미어 프로 CC 2021 - 쉽게 배워 제대로 써먹는 유튜브 영상 편집   
156  157                    실전 보고서 작성 기술 with 파워포인트, 워드, 한글   
198  199                                    처음 배우는 리액트 네이티브   

                           저자    출판사  발행년도           ISBN 세트 ISBN 부가기호     권  \
60         정지원,심수진,윤성우,김덕영 지음  한빛미디어  2021  9791162244029     NaN    1  2021   
70                        김도균  한빛미디어  2020  9791162243695     NaN  NaN   NaN   
88   정지원, 심수진, 윤성우, 김덕영 (지은이)  한빛미디어  2021  9791162244029     NaN  NaN   NaN   
156                    홍장표 지음  한빛미디어  2020  9791162243763     NaN  NaN   NaN   
198                    김범준 지음  한빛미디어  2021  9791162243879     NaN  NaN   NaN   

      주제분류번호  도서권수  대출건수        등록일자  
60   005.567     1     1  2021-03-15  
70    005.74     1     1  2021-03-15  
88       NaN     0     0  2021-03-15  
156      NaN     1     0  2021-03-12  
198      NaN     1     0  2021-03-12

In [19]:
#loc 메서드에 불리언 배열 넣으면 출판사가 '한빛미디어'인 모든 행 선택해서 같은 결과 나옴
ns_book2 = ns_book.loc[selected_rows]
ns_book2.head()

번호                                                도서명  \
60    61       (맛있는 디자인)프리미어 프로 CC: 쉽게 배워 제대로 써먹는 유튜브 영상 편집   
70    71         처음 배우는 애저 (Azure Portal로 배우는 애저 도입부터 활용까지)   
88    89  맛있는 디자인 프리미어 프로 CC 2021 - 쉽게 배워 제대로 써먹는 유튜브 영상 편집   
156  157                    실전 보고서 작성 기술 with 파워포인트, 워드, 한글   
198  199                                    처음 배우는 리액트 네이티브   

                           저자    출판사  발행년도           ISBN 세트 ISBN 부가기호     권  \
60         정지원,심수진,윤성우,김덕영 지음  한빛미디어  2021  9791162244029     NaN    1  2021   
70                        김도균  한빛미디어  2020  9791162243695     NaN  NaN   NaN   
88   정지원, 심수진, 윤성우, 김덕영 (지은이)  한빛미디어  2021  9791162244029     NaN  NaN   NaN   
156                    홍장표 지음  한빛미디어  2020  9791162243763     NaN  NaN   NaN   
198                    김범준 지음  한빛미디어  2021  9791162243879     NaN  NaN   NaN   

      주제분류번호  도서권수  대출건수        등록일자  
60   005.567     1     1  2021-03-15  
70    005.74     1     1  2021-03-15  
88       NaN     0     0  2021-03-15  
156      NaN     1     0  2021-03-12  
198      NaN     1     0  2021-03-12

### 대출권수 1,000 이하인 행 모두 삭제하기
판다스 이용자가 주로 사용하는 방법: 불리언 배열을 변수에 저장하는게 아니고(앞에서 selected_rows), [ ]에 불리언 배열 만드는 조건을 바로 넣기


In [20]:
ns_book2 = ns_book[ns_book['대출건수']>1000]
ns_book2.head()

번호                                      도서명                 저자  \
94781    94782  사피엔스 :유인원에서 사이보그까지, 인간 역사의 대담하고 위대한 질문   유발 하라리 지음 ;조현욱 옮김   
346944  346945                         해커스 토익:Listening       David Cho 지음   

             출판사  발행년도           ISBN 세트 ISBN 부가기호    권  주제분류번호  도서권수  대출건수  \
94781        김영사  2016  9788934972464     NaN  NaN  NaN     909    30  1468   
346944  해커스어학연구소  2005  9788990700148     NaN    1  NaN  740.77    29  1065   

              등록일자  
94781   2016-04-22  
346944  2005-02-01

# 중복된 행 찾기
duplicated() 메서드: 자동으로 중복된 행 찾음<br>
중복된 행 중 처음 행 제외한 나머지 행=True <br>
중복되지 않은 나머지 모든 행=False로 표시한 불리언 배열 반환

In [21]:
#sum함수를 사용하면 True=1로 인식->duplicated()메서드로 만든 불리언 배열의 중복된 행의 개수 세기
sum(ns_book.duplicated())

0

근데 모든 행이 고유의 '번호' 가지고 있어서 중복0개일 수밖에..<br> '도서명','저자'이런 **일부 열 기준으로 중복** 찾아보자<br>
subset매개변수에 기준열을 나열


In [22]:
sum(ns_book.duplicated(subset=['도서명','저자','ISBN']))

22096

# 여기 다시 찾아보며 공부
keep 매개변수는 어떤 중복값을 유지할지를 설정<br> false는 중복되지 않은 행들의 불리언값 반환<br>
 keep=False : 중복된 행을 모두 True로 표시한 불리언 배열 반환<br>
 keep='first' 사용 시 처음 나타난 값만 유지<br>
 keep='last'로 설정 시 마지막 나타난 값만 유지


In [23]:
dup_rows = ns_book.duplicated(subset=['도서명','저자','ISBN'], keep=False)
#얘는 아마 중복된 값을 반환하는 것 같지..?
ns_book3 = ns_book[dup_rows]
ns_book3.head()

번호                  도서명                 저자           출판사  발행년도  \
109  110                  파친코     이민진 지음 ;이미정 옮김          문학사상  2018   
110  111                  파친코     이민진 지음 ;이미정 옮김          문학사상  2018   
111  112  보건교사 안은영 :정세랑 장편소설            지은이: 정세랑           민음사  2021   
112  113  보건교사 안은영 :정세랑 장편소설            지은이: 정세랑           민음사  2021   
113  114                 스토너   존 윌리엄스 지음 ;김승욱 옮김  RHK(알에이치코리아)  2021   

              ISBN        세트 ISBN 부가기호    권 주제분류번호  도서권수  대출건수        등록일자  
109  9788970129815  9788970129808    0    1  843.6     1     0  2021-03-12  
110  9788970129822  9788970129808    0    2  843.6     1     0  2021-03-12  
111  9788937479953            NaN    0  NaN  813.7     1     0  2021-03-12  
112  9788937479953            NaN    0  NaN  813.7     1     1  2021-03-12  
113  9788925538297            NaN    0  NaN  843.5     1     0  2021-03-12

중복되는 행 그냥 지우는 법: drop_duplicates()메서드 이용. subset, keep 매개변수, 현재 데이터프레임을 수정하기 위한 inplace 매개변수가 있음

# 그룹별로 모으기
'혼공분석' 책 두 권의 대출권수 합쳐서 한 줄에 표시하기<br>
groupby()메서드 <br>
by매개변수에는 행을 합칠 때 기준이 되는 열을 지정

In [24]:
#ns_book 데이터프레임에서 그룹으로 묶을 기준 열(4개)과 '대출건수'열만 선택해서 사용
count_df = ns_book[['도서명','저자','ISBN','권','대출건수']]

groupby()메서드는 기본적으로 NaN가 포함된 열은 자동으로 삭제<br>
-> 그러면 대출건수 합칠 때 누락되는 경우 발생<br>
->dropna매개변수=False로 지정하면 NaN 있는 행도 포함할 수 있음

In [25]:
 group_df=count_df.groupby(by=['도서명','저자','ISBN','권','대출건수'])
 loan_count=group_df.sum() #대출건수는 하나로 합쳐야하니까 sum()메서드 사용

In [26]:
#위 코드 한방에 합친 코드
loan_count = count_df.groupby(by=['도서명','저자','ISBN','권'],dropna=False).sum()
loan_count.head()

대출건수
도서명                                             저자                       ISBN          권        
 (꼭 필요한 것부터 쉽게 배우는) 자신만만 블로그 차근차근 배우기           김상현 지음                   9788955025637 NaN    38
 (맨처음 배우는) 세상의 직업                               엘레오노라 바르소티 글 ·그림 ;김태은 옮김 9788992924146 NaN    10
 (영잘원 리스닝과 패턴 영어의 절묘한 만남으로 태어난 ) 리스닝 ABC : 입문편  JD Kim 지음                9788993466089 NaN     4
 (즉석에서 바로바로 활용하는) 일상생활 베트남어 첫걸음                 FL4U컨텐츠 지음               9788971728000 NaN     3
 ,에게                                            이기린(이진희)                 9791196137014 NaN     0

## 대출건수 원본 데이터에 업데이트하기
###1. 원본 데이터에 중복된 데이터 정리하기<br>
 1.1. duplicated()메서드로 중복된 행 True로 표시한 불리언 배열 만들기<br>
 1.2. 앞서 구한 불리언 배열을 반전시켜 중복되지 않은 고유한 행을 True로 만들기<br>:판다스의 ~연산자 사용<br>
 1.3. 앞서 2번에서 구한 불리언 배열을 사용해 원본 배열에서 고유한 행만 선택<br>:copy()메서드로 ns_book3 데이터프레임 복사본 만들기<br>

In [27]:
dup_rows = ns_book.duplicated(subset=['도서명','저자','ISBN','권']) #중복된 행 True로 표시
unique_rows=~dup_rows
ns_book3=ns_book[unique_rows].copy() #고유한 행만 선택

In [28]:
#중복된 행 있는지 duplicated() 메서드로 확인
sum(ns_book3.duplicated(subset=['도서명','저자','ISBN','권']))

0

set_index()메서드 = 지정한 열을 인덱스로 설정
inplace 매개변수=True ->새로운 데이터프레임 반환 x. ns_book3 데이터프레임 수정o

In [29]:
#ns_book3의 인덱스를 loan_count 데이터프레임의 인덱스와 동일하게 만들기
ns_book3.set_index(['도서명','저자','ISBN','권'], inplace=True)
ns_book3.head()

번호    출판사  발행년도  \
도서명                   저자               ISBN          권                      
인공지능과 흙               김동훈 지음           9788937444319 NaN   1    민음사  2021   
가짜 행복 권하는 사회          김태형 지음           9791190123969 NaN   2   갈매나무  2021   
나도 한 문장 잘 쓰면 바랄 게 없겠네 김선영 지음           9788968332982 NaN   3   블랙피쉬  2021   
예루살렘 해변               이도 게펜 지음, 임재희 옮김 9788970759906 NaN   4  문학세계사  2021   
김성곤의 중국한시기행 : 장강·황하 편 김성곤 지음           9788934990833 NaN   5    김영사  2021   

                                                         세트 ISBN 부가기호 주제분류번호  \
도서명                   저자               ISBN          권                         
인공지능과 흙               김동훈 지음           9788937444319 NaN     NaN  NaN    NaN   
가짜 행복 권하는 사회          김태형 지음           9791190123969 NaN     NaN  NaN    NaN   
나도 한 문장 잘 쓰면 바랄 게 없겠네 김선영 지음           9788968332982 NaN     NaN  NaN    NaN   
예루살렘 해변               이도 게펜 지음, 임재희 옮김 9788970759906 NaN     NaN  NaN    NaN   
김성곤의 중국한시기행 : 장강·황하 편 김성곤 지음           9788934990833 NaN     NaN  NaN    NaN   

                                                          도서권수  대출건수  \
도서명                   저자               ISBN          권                 
인공지능과 흙               김동훈 지음           9788937444319 NaN     1     0   
가짜 행복 권하는 사회          김태형 지음           9791190123969 NaN     1     0   
나도 한 문장 잘 쓰면 바랄 게 없겠네 김선영 지음           9788968332982 NaN     1     0   
예루살렘 해변               이도 게펜 지음, 임재희 옮김 9788970759906 NaN     1     0   
김성곤의 중국한시기행 : 장강·황하 편 김성곤 지음           9788934990833 NaN     1     0   

                                                                등록일자  
도서명                   저자               ISBN          권                
인공지능과 흙               김동훈 지음           9788937444319 NaN  2021-03-19  
가짜 행복 권하는 사회          김태형 지음           9791190123969 NaN  2021-03-19  
나도 한 문장 잘 쓰면 바랄 게 없겠네 김선영 지음           9788968332982 NaN  2021-03-19  
예루살렘 해변               이도 게펜 지음, 임재희 옮김 9788970759906 NaN  2021-03-19  
김성곤의 중국한시기행 : 장강·황하 편 김성곤 지음           9788934990833 NaN  2021-03-19

ns_book3의 '대출권수'열을 loan_count 데이터프레임의 '대출건수' 열로 업데이트하기<br>
update()메서드 : 다른 데이터프레임을 사용해 원본 데이터프레임 값을 업데이트

In [30]:
ns_book3.update(loan_count)
ns_book3.head()

번호    출판사  발행년도  \
도서명                   저자               ISBN          권                      
인공지능과 흙               김동훈 지음           9788937444319 NaN   1    민음사  2021   
가짜 행복 권하는 사회          김태형 지음           9791190123969 NaN   2   갈매나무  2021   
나도 한 문장 잘 쓰면 바랄 게 없겠네 김선영 지음           9788968332982 NaN   3   블랙피쉬  2021   
예루살렘 해변               이도 게펜 지음, 임재희 옮김 9788970759906 NaN   4  문학세계사  2021   
김성곤의 중국한시기행 : 장강·황하 편 김성곤 지음           9788934990833 NaN   5    김영사  2021   

                                                         세트 ISBN 부가기호 주제분류번호  \
도서명                   저자               ISBN          권                         
인공지능과 흙               김동훈 지음           9788937444319 NaN     NaN  NaN    NaN   
가짜 행복 권하는 사회          김태형 지음           9791190123969 NaN     NaN  NaN    NaN   
나도 한 문장 잘 쓰면 바랄 게 없겠네 김선영 지음           9788968332982 NaN     NaN  NaN    NaN   
예루살렘 해변               이도 게펜 지음, 임재희 옮김 9788970759906 NaN     NaN  NaN    NaN   
김성곤의 중국한시기행 : 장강·황하 편 김성곤 지음           9788934990833 NaN     NaN  NaN    NaN   

                                                          도서권수  대출건수  \
도서명                   저자               ISBN          권                 
인공지능과 흙               김동훈 지음           9788937444319 NaN     1   0.0   
가짜 행복 권하는 사회          김태형 지음           9791190123969 NaN     1   0.0   
나도 한 문장 잘 쓰면 바랄 게 없겠네 김선영 지음           9788968332982 NaN     1   0.0   
예루살렘 해변               이도 게펜 지음, 임재희 옮김 9788970759906 NaN     1   0.0   
김성곤의 중국한시기행 : 장강·황하 편 김성곤 지음           9788934990833 NaN     1   0.0   

                                                                등록일자  
도서명                   저자               ISBN          권                
인공지능과 흙               김동훈 지음           9788937444319 NaN  2021-03-19  
가짜 행복 권하는 사회          김태형 지음           9791190123969 NaN  2021-03-19  
나도 한 문장 잘 쓰면 바랄 게 없겠네 김선영 지음           9788968332982 NaN  2021-03-19  
예루살렘 해변               이도 게펜 지음, 임재희 옮김 9788970759906 NaN  2021-03-19  
김성곤의 중국한시기행 : 장강·황하 편 김성곤 지음           9788934990833 NaN  2021-03-19

업데이트 완료되었으니, 인덱스 열 해제 : reset_index()메서드

In [31]:
ns_book4 = ns_book3.reset_index()
ns_book4.head()

도서명                저자           ISBN    권  번호    출판사  \
0                인공지능과 흙            김동훈 지음  9788937444319  NaN   1    민음사   
1           가짜 행복 권하는 사회            김태형 지음  9791190123969  NaN   2   갈매나무   
2  나도 한 문장 잘 쓰면 바랄 게 없겠네            김선영 지음  9788968332982  NaN   3   블랙피쉬   
3                예루살렘 해변  이도 게펜 지음, 임재희 옮김  9788970759906  NaN   4  문학세계사   
4  김성곤의 중국한시기행 : 장강·황하 편            김성곤 지음  9788934990833  NaN   5    김영사   

   발행년도 세트 ISBN 부가기호 주제분류번호  도서권수  대출건수        등록일자  
0  2021     NaN  NaN    NaN     1   0.0  2021-03-19  
1  2021     NaN  NaN    NaN     1   0.0  2021-03-19  
2  2021     NaN  NaN    NaN     1   0.0  2021-03-19  
3  2021     NaN  NaN    NaN     1   0.0  2021-03-19  
4  2021     NaN  NaN    NaN     1   0.0  2021-03-19

### 대출건수 잘 합쳐졌나 확인

In [32]:
#원본 데이터프레임 ns_book에서 대출건수 100회 이상인 책 개수 세기
sum(ns_book['대출건수']>100)

2311

In [34]:
#새로 만든 데이터프레임 ns_book4에서 대출건수 100회 이상인 책 개수 세기
#중복도서 합쳐져서 늘어난 것 확인할 수 있음
sum(ns_book4['대출건수']>100)

2550

### 데이터프레임 열 순서 원래대로 돌리기
여러 개의 열 사용해 인덱스 만들었다 해제해서, 초기 열과 ns_book4의 열 순서가 달라졌음<br>
-> 이후 분석과정에서 혼돈 막기 위해 원래대로 돌리자<br>
-> [ ] 연산자에 원하는 열 이름 순서대로 전달

In [35]:
ns_book4 = ns_book4[ns_book.columns]
ns_book4.head()

번호                    도서명                저자    출판사  발행년도           ISBN  \
0   1                인공지능과 흙            김동훈 지음    민음사  2021  9788937444319   
1   2           가짜 행복 권하는 사회            김태형 지음   갈매나무  2021  9791190123969   
2   3  나도 한 문장 잘 쓰면 바랄 게 없겠네            김선영 지음   블랙피쉬  2021  9788968332982   
3   4                예루살렘 해변  이도 게펜 지음, 임재희 옮김  문학세계사  2021  9788970759906   
4   5  김성곤의 중국한시기행 : 장강·황하 편            김성곤 지음    김영사  2021  9788934990833   

  세트 ISBN 부가기호    권 주제분류번호  도서권수  대출건수        등록일자  
0     NaN  NaN  NaN    NaN     1   0.0  2021-03-19  
1     NaN  NaN  NaN    NaN     1   0.0  2021-03-19  
2     NaN  NaN  NaN    NaN     1   0.0  2021-03-19  
3     NaN  NaN  NaN    NaN     1   0.0  2021-03-19  
4     NaN  NaN  NaN    NaN     1   0.0  2021-03-19

In [36]:
#ns_book4 데이터프레임 저장해두기
ns_book4.to_csv('ns_book4.csv',index=False)

#정리: 일괄 처리 함수 만들기
**열 삭제 방법**: loc메서드에 슬라이싱or불리언 배열 적용 / drop()메서드<br>
**행 삭제 방법**: [ ] 연산자와 슬라이싱, 불리언 배열 (loc도가능하지만 잘안씀)<br>
**중복된 행 찾기**: groupby(), sum()메서드 사용해 중복된 행 대출건수 합치고 / update()메서드로 원본 데이터프레임에 업데이트. 인덱스 기준 업데이트 위해 set_index(), reset()메서드 사용

In [87]:
#일괄처리함수 만들기
def data_cleaning(filename):
  """
  남산 도서관 장서 CSV 데이터 전처리 함수
  :param filename: CSV 파일 이름
  """
  #파일을 데이터프레임으로 읽기
  ns_df = pd.read_csv(filename, low_memory=False)
  #NaN인 열 삭제하기
  ns_book = ns_df.dropna(axis=1, how='all')
  #대출건수 합치기 위해 필요한 행만 추출하여 count_df데이터프레임 만들기
  count_df=ns_book[['도서명','저자','ISBN','권','대출건수']]
  #도서명, 저자, ISBN, 권을 기준으로 대출권수 groupby
  loan_count = count_df.groupby(by=['도서명','저자','ISBN','권'],dropna=False).sum()
  dup_rows=ns_book.duplicated(subset=['도서명','저자','ISBN','권'])
  unique_rows=~dup_rows
  ns_book3=ns_book[unique_rows].copy()
  #도서명, 저자, isbn, 권을 인덱스로 설정
  ns_book3.set_index(['도서명','저자','ISBN','권'],inplace=True)
  #dloan_count에 저장된 누적 대출건수 업데이트
  ns_book3.update(loan_count)
  #인덱스 재설정
  ns_book4 = ns_book3.reset_index()
  #원본 데이터프레임의 열 순서로 변경
  ns_book4 = ns_book4[ns_book.columns]
  return ns_book4

In [88]:
#원본데이터인 ns_202104.csv 파일로 함수 제대로 동작하는지 확인
#equals() 메서드 사용
new_ns_book4 = data_cleaning('ns_202104.csv')
ns_book4.equals(new_ns_book4)


False

#3-2 잘못된 데이터 수정하기
누락된 값, 잘못된 값 다루기<br>
NaN 값 확인하기

In [89]:
import gdown

gdown.download('https://bit.ly/3GisL6J', 'ns_book4.csv', quiet=False)

Downloading...
From: https://bit.ly/3GisL6J
To: /content/ns_book4.csv
100%|██████████| 55.5M/55.5M [00:00<00:00, 79.0MB/s]


'ns_book4.csv'

In [90]:
import pandas as pd

ns_book4 = pd.read_csv('ns_book4.csv', low_memory=False)
ns_book4.head()

번호                    도서명                저자    출판사  발행년도           ISBN  \
0   1                인공지능과 흙            김동훈 지음    민음사  2021  9788937444319   
1   2           가짜 행복 권하는 사회            김태형 지음   갈매나무  2021  9791190123969   
2   3  나도 한 문장 잘 쓰면 바랄 게 없겠네            김선영 지음   블랙피쉬  2021  9788968332982   
3   4                예루살렘 해변  이도 게펜 지음, 임재희 옮김  문학세계사  2021  9788970759906   
4   5  김성곤의 중국한시기행 : 장강·황하 편            김성곤 지음    김영사  2021  9788934990833   

  세트 ISBN 부가기호    권 주제분류번호  도서권수  대출건수        등록일자  
0     NaN  NaN  NaN    NaN     1   0.0  2021-03-19  
1     NaN  NaN  NaN    NaN     1   0.0  2021-03-19  
2     NaN  NaN  NaN    NaN     1   0.0  2021-03-19  
3     NaN  NaN  NaN    NaN     1   0.0  2021-03-19  
4     NaN  NaN  NaN    NaN     1   0.0  2021-03-19

## 누락된 값 처리하기
1. info()메서드로 열마다 NaN이 아닌 값 개수 확인

In [91]:
ns_book4.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 384591 entries, 0 to 384590
Data columns (total 13 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   번호       384591 non-null  int64  
 1   도서명      384188 non-null  object 
 2   저자       384393 non-null  object 
 3   출판사      379950 non-null  object 
 4   발행년도     384577 non-null  object 
 5   ISBN     384591 non-null  object 
 6   세트 ISBN  56576 non-null   object 
 7   부가기호     310386 non-null  object 
 8   권        63378 non-null   object 
 9   주제분류번호   364727 non-null  object 
 10  도서권수     384591 non-null  int64  
 11  대출건수     384591 non-null  float64
 12  등록일자     384591 non-null  object 
dtypes: float64(1), int64(2), object(10)
memory usage: 38.1+ MB


전체 행 개수 384591<br>
열 개수 13<br>
int64/float64/object(문자열,혼합형): 열 데이터 타입<br>
Non-Null 밑에: 누락된 값이 없는 행 개수<br>
=>누락된 값이 있음을 확인 가능


### 누락된 값 처리하기
1. 누락 개수 확인
2. 누락된 값 표시하는 방법
3. 누락된 값 다른 값으로 바꾸기

*1. 누락 개수 확인*

In [92]:
#누락 개수 확인: isan()메서드 -> 각 행의 공란 여부를 불리언 배열로 반환
#sum() 메서드로 True 개수로 비어 있는 행 개수 얻기

ns_book4.isna().sum()

번호              0
도서명           403
저자            198
출판사          4641
발행년도           14
ISBN            0
세트 ISBN    328015
부가기호        74205
권          321213
주제분류번호      19864
도서권수            0
대출건수            0
등록일자            0
dtype: int64

*2. 누락된 값 표시*

In [93]:
#임의로 누락된 값 만들고 NaN로 표시
#판다스 데이터프레임에서는, 정수를 저장하는 열에 파이썬의 None 입력시 누락된 값으로 인식
#ns_book4 데이터프레임의 '도서권수' 열의 첫 번째 행 값을 None으로 바꾼 후, '도서권수' 열에 isna()메서드 적용하여 누락된 값 카운트

ns_book4.loc[0, '도서권수'] = None
ns_book4['도서권수'].isna().sum()

1

In [94]:
ns_book4.head(2)

번호           도서명      저자   출판사  발행년도           ISBN 세트 ISBN 부가기호    권  \
0   1       인공지능과 흙  김동훈 지음   민음사  2021  9788937444319     NaN  NaN  NaN   
1   2  가짜 행복 권하는 사회  김태형 지음  갈매나무  2021  9791190123969     NaN  NaN  NaN   

  주제분류번호  도서권수  대출건수        등록일자  
0    NaN   NaN   0.0  2021-03-19  
1    NaN   1.0   0.0  2021-03-19

도서권수 두 번째 값 1.0 된 건, 첫 번째 값 NaN(실수값으로 저장됨)로 바뀌면서 float64로 바뀌었기 때문

*3. 누락된 값 다른 값으로 바꿔보기*

데이터 타입 지정 : astype()메서드<br>
매개변수 {열 이름:데이터 타입} 딕셔너리 형식으로 전달

In [95]:
#'도서건수' 실수형 다시 정수형으로 바꾸기
#이왕 하는거 '대출건수'도 데이터 타입 int32로 바꿔보기
ns_book4.loc[0, '도서권수'] = 1
ns_book4 = ns_book4.astype({'도서권수':'int32', '대출건수': 'int32'})
ns_book4.head(2)

번호           도서명      저자   출판사  발행년도           ISBN 세트 ISBN 부가기호    권  \
0   1       인공지능과 흙  김동훈 지음   민음사  2021  9788937444319     NaN  NaN  NaN   
1   2  가짜 행복 권하는 사회  김태형 지음  갈매나무  2021  9791190123969     NaN  NaN  NaN   

  주제분류번호  도서권수  대출건수        등록일자  
0    NaN     1     0  2021-03-19  
1    NaN     1     0  2021-03-19

In [96]:
#문자열 저장할 수 있는 object인 '부가기호' 열 첫 행에 None 입력해보기

ns_book4.loc[0, '부가기호'] = None
ns_book4.head(2)

번호           도서명      저자   출판사  발행년도           ISBN 세트 ISBN  부가기호    권  \
0   1       인공지능과 흙  김동훈 지음   민음사  2021  9788937444319     NaN  None  NaN   
1   2  가짜 행복 권하는 사회  김태형 지음  갈매나무  2021  9791190123969     NaN   NaN  NaN   

  주제분류번호  도서권수  대출건수        등록일자  
0    NaN     1     0  2021-03-19  
1    NaN     1     0  2021-03-19

In [97]:
#근데 판다스에는 NaN라는 값이 없고, 넘파이 패키지의 np.nan을 사용해야함.
#따라서 첫 행의 '부가기호' 열 값을 None에서 NaN로 표시하려면 np.nan을 사용
import numpy as np

ns_book4.loc[0, '부가기호'] = np.nan
ns_book4.head(2)

번호           도서명      저자   출판사  발행년도           ISBN 세트 ISBN 부가기호    권  \
0   1       인공지능과 흙  김동훈 지음   민음사  2021  9788937444319     NaN  NaN  NaN   
1   2  가짜 행복 권하는 사회  김태형 지음  갈매나무  2021  9791190123969     NaN  NaN  NaN   

  주제분류번호  도서권수  대출건수        등록일자  
0    NaN     1     0  2021-03-19  
1    NaN     1     0  2021-03-19

## 누락된 값 바꾸기
방법1. loc, fillna() 메서드<br>
loc메서드: 누락된 값을 원하는 값으로 바꿀 수 있음<br>
-> 누락된 값을 가리키는 불리언 배열 필요<br>
-> isna()메서드 사용

In [98]:
#누락된 값 NaN을 빈 문자열('')로 바꾸기
#'세트 ISBN'열의 NaN을 가리키는 set_isbn_na_rows 불리언 배열 만들고 loc메서드에 전달
set_isbn_na_rows = ns_book4['세트 ISBN'].isna()
ns_book4.loc[set_isbn_na_rows, '세트 ISBN'] = ''

ns_book4['세트 ISBN'].isna().sum() #누락된 값 개수 카운트

0

fillna()메서드에 원하는 값 전달하면 NaN 대체 가능


In [99]:
#ns_book4의 모든 NaN값을 '없음' 문자열로 변환
ns_book4.fillna('없음').isna().sum()
#모두 '없음'으로 바뀌어서, NaN개수를 세면 0 출력됨.

번호         0
도서명        0
저자         0
출판사        0
발행년도       0
ISBN       0
세트 ISBN    0
부가기호       0
권          0
주제분류번호     0
도서권수       0
대출건수       0
등록일자       0
dtype: int64

In [100]:
#특정 열만 선택해서 NaN 바꾸기
ns_book4['부가기호'].fillna('없음').isna().sum()

0

특정 열 선택-fillna()메서드: 열 이름 없이 개수만 나오는 판다스 시리즈 객체로 반환<br>
'부가기호' 열의 NaN을 바꾸면서 전체 데이터프레임 반환하려면, 열 이름과 바꾸려는 값으로 이루어진 딕셔너리를 전달

In [101]:
ns_book4.fillna({'부가기호':'없음'}).isna().sum()

번호              0
도서명           403
저자            198
출판사          4641
발행년도           14
ISBN            0
세트 ISBN         0
부가기호            0
권          321213
주제분류번호      19864
도서권수            0
대출건수            0
등록일자            0
dtype: int64

####방법2. replace 메서드<br>
**바꾸려는 값이 하나일 때**: replace(원래 값, 새로운 값)

In [102]:
ns_book4.replace(np.nan, '없음').isna().sum()

번호         0
도서명        0
저자         0
출판사        0
발행년도       0
ISBN       0
세트 ISBN    0
부가기호       0
권          0
주제분류번호     0
도서권수       0
대출건수       0
등록일자       0
dtype: int64

**바꾸려는 값이 여러 개일 때**: replace([원래 값1, 원래 값2],[새로운 값1, 새로운 값2])

In [103]:
#NaN->'없음' , '2021'->'21'
ns_book4.replace([np.nan, '2021'], ['없음', '21']).head(2)

번호           도서명      저자   출판사 발행년도           ISBN 세트 ISBN 부가기호   권 주제분류번호  \
0   1       인공지능과 흙  김동훈 지음   민음사   21  9788937444319           없음  없음     없음   
1   2  가짜 행복 권하는 사회  김태형 지음  갈매나무   21  9791190123969           없음  없음     없음   

   도서권수  대출건수        등록일자  
0     1     0  2021-03-19  
1     1     0  2021-03-19

**열 마다 다른 값으로 바꿀 때**: replace({열 이름:원래 값}, 새로운 값)

In [104]:
#'부가기호' 열의 NaN->'없음'

ns_book4.replace({'부가기호': np.nan}, '없음').head(2)

번호           도서명      저자   출판사  발행년도           ISBN 세트 ISBN 부가기호    권  \
0   1       인공지능과 흙  김동훈 지음   민음사  2021  9788937444319           없음  NaN   
1   2  가짜 행복 권하는 사회  김태형 지음  갈매나무  2021  9791190123969           없음  NaN   

  주제분류번호  도서권수  대출건수        등록일자  
0    NaN     1     0  2021-03-19  
1    NaN     1     0  2021-03-19

In [105]:
#'부가기호'열의 NaN->'없음', '발행년도'열의 '2021'->'21'
ns_book4.replace({'부가기호': {np.nan: '없음'}, '발행년도': {'2021': '21'}}).head(2)

번호           도서명      저자   출판사 발행년도           ISBN 세트 ISBN 부가기호    권  \
0   1       인공지능과 흙  김동훈 지음   민음사   21  9788937444319           없음  NaN   
1   2  가짜 행복 권하는 사회  김태형 지음  갈매나무   21  9791190123969           없음  NaN   

  주제분류번호  도서권수  대출건수        등록일자  
0    NaN     1     0  2021-03-19  
1    NaN     1     0  2021-03-19

## 정규 표현식(정규식)
ex. 2021->21 , 2018->18 이렇게 replace로 다 일일이 치기 힘들어. -> 정규표현식 활용
> 정규 표현식에서 숫자 = \d<br>
네 자리 연도 \d\d\d\d<br>
뒤에 두 자리 그룹으로 \d\d(\d\d) <br>
-> 패턴에 맞는 문자열 찾은 뒤, 첫 번째 그룹에 해당하는 뒷자리 연도 두 개 추출<br>


*   \1, \2 : 몇 번째 등장 그룹?



In [106]:
#'발행년도'열의 값을 정규 표현식으로 두 자리 연도로 바꾸기
#regex 매개변수: 정규 표현식을 사용한다는 의미. replace()와 유사
ns_book4.replace({'발행년도': {r'\d\d(\d\d)': r'\1'}}, regex=True)[100:102]
#패턴 안의 첫 번째 그룹으로 연도를 바꾼다
#r문자=다른 문자열과 정규식을 구분하기 위해 접두사처럼 붙임
#중괄호로 반복되는 정규 표현식 표현 가능 \d{2}(\d{2})=\d\d(\d\d)
#=ns_book4.replace({'발행년도': {r'\d{2}(\d{2})': r'\1'}}, regex=True)[100:102]

번호                                        도서명  \
100  101                  No라고 말할 줄 아는 남편과 아내 - 개정판   
101  102  D2C 레볼루션 - 스타트업부터 글로벌 기업까지, 마켓 체인저의 필수 전략   

                                   저자   출판사 발행년도           ISBN 세트 ISBN 부가기호  \
100  헨리 클라우드, 존 타운센드 (지은이), 김진웅 (옮긴이)  좋은씨앗   18  9788958743019          NaN   
101        로런스 인그래시아 (지은이), 안기순 (옮긴이)    부키   21  9788960518483          NaN   

       권 주제분류번호  도서권수  대출건수        등록일자  
100  NaN  234.9     1     1  2021-03-15  
101  NaN  325.1     1     0  2021-03-15

문자찾기: 마침표(.)<br>
**.** : 모든 문자에 대응하는 정규 표현식 문자<br>
이름은 연도처럼 몇 개 문자 반복될지 예측 불가->**\***사용<br>
저자 항목에 (지은이), (옮긴이)는 삭제<br>
=>정규표현식에서 ()는 그룹 나타내는 데 쓰임->일반 문자로 인식하기 위해서는 앞에 '\\'


In [107]:
ns_book4.replace({'저자': {r'(.*)\s\(지은이\)(.*)\s\(옮긴이\)': r'\1\2'},
                  '발행년도': {r'\d{2}(\d{2})': r'\1'}}, regex=True)[100:102]

번호                                        도서명                    저자  \
100  101                  No라고 말할 줄 아는 남편과 아내 - 개정판  헨리 클라우드, 존 타운센드, 김진웅   
101  102  D2C 레볼루션 - 스타트업부터 글로벌 기업까지, 마켓 체인저의 필수 전략        로런스 인그래시아, 안기순   

      출판사 발행년도           ISBN 세트 ISBN 부가기호    권 주제분류번호  도서권수  대출건수        등록일자  
100  좋은씨앗   18  9788958743019          NaN  NaN  234.9     1     1  2021-03-15  
101    부키   21  9788960518483          NaN  NaN  325.1     1     0  2021-03-15

## 잘못된 값 바꾸기

In [108]:
#astype()로 '발행년도'열을 int32로 바꾸기
ns_book4.astype({'발행년도':'int32'})
#오류발생. '1998.'이라는 연도를 변환할 수 없어서. =숫자+다른 문자가 같이 있어서 오류난거임

ValueError: invalid literal for int() with base 10: '1988.'

In [ ]:
#contains()로 문자열 패턴 포함여부 확인
#발행년도에 1988 포함된 행 개수 세기
ns_book4['발행년도'].str.contains('1988').sum()

모든 문자에 대응하는 표현은 \\D

In [ ]:
#숫자가 아닌 문자를 포함하는 모든 행 찾기
#contains매개변수 na=True : 연도가 누락된 행을 True로 표시
#->누락값을자동으로 np.nan으로 채워서 invaild_number 배열을 인덱싱에 사용할 수 없기 때문
invalid_number = ns_book4['발행년도'].str.contains('\D', na=True)
print(invalid_number.sum())
ns_book4[invalid_number].head()

In [109]:
#정규표현식을 통해 연도 앞뒤의 문자 제거
ns_book5 = ns_book4.replace({'발행년도':r'.*(\d{4}).*'}, r'\1', regex=True)
ns_book5[invalid_number].head()

번호                       도서명  \
19138  19565                      단국강토   
19227  19736                    삼성의 역사   
26097  26812                  배고픈 애벌레    
29817  30586  (The) Sopranos sessions    
29940  30709                    다음엔 너야   

                                                      저자                 출판사  \
19138                                              홍태수 저             매일경제신문사   
19227                                              송부웅 撰                  삼양   
26097                                  에릭 칼 글·그림 ;이희재 옮김               더큰컴퍼니   
29817  Matt Zoller Seitz,$eAlan Sepinwall,$eLaura Lip...  Harry N Abrams Inc   
29940                         에른스트 얀들 글;노르만 융에 그림;박상순 옮김                 비룡소   

       발행년도           ISBN        세트 ISBN 부가기호    권  주제분류번호  도서권수  대출건수  \
19138  1988  9788974420031                 NaN  NaN   511.1     1     1   
19227  2001  9788985464369                   0  NaN  911.02     1     1   
26097  2019  9788959514083                 NaN  NaN     843     1     0   
29817  2019  9781419734946                 NaN  NaN  326.76     1     0   
29940  2018  9788949110646  9788949110004    7  NaN     853     1     9   

             등록일자  
19138  2019-12-19  
19227  2019-12-06  
26097  2019-08-12  
29817  2019-06-13  
29940  2019-06-04

In [110]:
#문자 들어간 행의 개수와 데이터 재확인
unkown_year = ns_book5['발행년도'].str.contains('\D', na=True)
print(unkown_year.sum())
ns_book5[unkown_year].head()

67


번호                도서명        저자      출판사     발행년도           ISBN  \
30838    31616       본격 한중일 세계사 5  굽시니스트 지음   위즈덤하우스      NaN  9791190065092   
39130    40141     정책금융의 현황과 발전과제   정책금융연구회   한국산업은행      NaN  9788992784108   
39256    40268  서울지역 유적 발굴조사 총서 3   서울역사박물관  서울역사박물관      NaN  9791186324714   
76836    81202            흰머리 큰줄기    한호진 지음    秀文出版社  [발행년불명]  9788973010769   
150543  160436          (속) 경제학사     박장환 지음      NaN   [20--]  9788994339207   

              세트 ISBN 부가기호    권 주제분류번호  도서권수  대출건수        등록일자  
30838                  NaN  NaN    NaN     0     0  2019-05-28  
39130                  NaN  NaN  327.1     1     0  2019-01-22  
39256   9791186324431  NaN  NaN    NaN     0     0  2019-01-22  
76836                    0  NaN  699.1     1     1  2016-11-10  
150543                   1  NaN  320.9     1     2  2012-11-19

In [111]:
#NaN또는 네 자리 숫자가 아닌 값->임의로 -1 값으로 바꾼 후, astype()메서드로 '발행년도' 열의 데이터 타입을 정수형인 int32로 변환
ns_book5.loc[unkown_year, '발행년도'] = '-1'
ns_book5 = ns_book5.astype({'발행년도': 'int32'})

In [112]:
#gt메서드: 전달받은 값보다 큰 값을 찾음
#연도가 4000년이 넘는 경우 찾기
ns_book5['발행년도'].gt(4000).sum()

131

In [113]:
#단군기원을 사용한 연도이기 때문에, 2333을 빼서 서기로 바꾼 다음 4000년이 넘는 도서 유무를 확인
dangun_yy_rows = ns_book5['발행년도'].gt(4000)
ns_book5.loc[dangun_yy_rows, '발행년도'] = ns_book5.loc[dangun_yy_rows, '발행년도'] - 2333

dangun_year = ns_book5['발행년도'].gt(4000)
print(dangun_year.sum())
ns_book5[dangun_year].head(2)

13


번호                                         도서명               저자  \
222858  234268                                  Modern art  [computer file]   
270269  282852  현대 영어학=Linguistic theory in modern english              이재영   

            출판사  발행년도           ISBN 세트 ISBN 부가기호    권   주제분류번호  도서권수  대출건수  \
222858  GROLIER  7611  9780717233243          NaN  NaN  609.205     1     1   
270269     열림기획  7634  9788986705072            1  NaN    740.1     1     6   

              등록일자  
222858  2009-05-07  
270269  2007-04-14

In [114]:
#아직도 남은 책. -1로 표시
ns_book5.loc[dangun_year, '발행년도'] = -1

In [115]:
#연도가 작은 값 확인(0~1900년 이전)
#gt:보다 큰 값 #lt:보다 작은 값
old_books = ns_book5['발행년도'].gt(0) & ns_book5['발행년도'].lt(1900)
ns_book5[old_books]

번호                                           도서명  \
61149    64093         고흐 씨, 시 읽어 줄까요 :내 마음을 알아주는 시와 그림의 만남    
70566    74347                                  정신 병리학의 문제들    
79550    84164  (최근 7년간) 중요 민법판례 :[2009년 1월~2016년 6월 중요 판례]    
147950  157759              (한·중·일) 밥상 문화 :대표음식으로 본 3국 문화비교    
194576  205407                                책으로 뒤덮인 집의 비밀    
287252  300283                                     (밝혀질)우리歷史   

                                   저자      출판사  발행년도           ISBN  \
61149                          이운진 지음      사계절   176  9788958284277   
70566        지은이: 지그문트 프로이트 ;옮긴이: 황보석     열린책들   151  9788932905181   
79550                          이광섭 편저      법학사   163  9788962898651   
147950                         김경은 지음      이가서   132  9788958643012   
194576  N.E. 보드 지음 ;피터 퍼거슨 그림 ;김지현 옮김     개암나무  1015  9788992844413   
287252                          吳在城 著  黎民族史硏究會  1607  9788986892130   

              세트 ISBN 부가기호    권  주제분류번호  도서권수  대출건수        등록일자  
61149                    0  NaN   811.7     1     6  2017-10-27  
70566   9788932905082    9   10   185.5     1     4  2017-04-26  
79550                    9  NaN     365     1    14  2016-09-27  
147950                   0  NaN  381.75     1    30  2013-02-19  
194576                   7  NaN     843     1    15  2010-08-18  
287252                   0  NaN     911     1     5  2006-07-06

In [116]:
#발행년도 아직도 이상한거 있으니, -1로 설정하고 전체 행 개수 확인

ns_book5.loc[old_books, '발행년도'] = -1
ns_book5['발행년도'].eq(-1).sum()

86

####다른 열의 잘못'누락된 값 채우기
도서명, 저자, 출판사, 발행년도 열 정리

In [117]:
#누락된 값 있거나, 발행년도 -1인 행 개수 확인
na_rows = ns_book5['도서명'].isna() | ns_book5['저자'].isna() \
          | ns_book5['출판사'].isna() | ns_book5['발행년도'].eq(-1)
print(na_rows.sum())
ns_book5[na_rows].head(2)

5268


번호          도서명          저자  출판사  발행년도           ISBN 세트 ISBN 부가기호    권  \
78    79   아산 정주영 레거시         김화진  NaN  2021  9788952129529          NaN  NaN   
265  278  골목의 시간을 그리다  정명섭.김효찬 지음  NaN  2021  9791191266054          NaN  NaN   

    주제분류번호  도서권수  대출건수        등록일자  
78     325     1     1  2021-03-15  
265    NaN     1     0  2021-03-12

In [118]:
#5268개의 값을 뷰티플수프로 채우기

import requests
from bs4 import BeautifulSoup

#yes24 크롬 개발자 도구로 도서명 태그정보 확인
def get_book_title(isbn):
    # Yes24 도서 검색 페이지 URL
    url = 'http://www.yes24.com/Product/Search?domain=BOOK&query={}'
    # URL에 ISBN을 넣어 HTML 가져오기
    r = requests.get(url.format(isbn))
    soup = BeautifulSoup(r.text, 'html.parser')   # HTML 파싱
    # 클래스 이름이 'gd_name'인 a 태그의 텍스트를 가져오기
    title = soup.find('a', attrs={'class':'gd_name'}) \
            .get_text()
    return title

In [119]:
#골목의 시간을 그리다 책의 isbn으로 테스트

get_book_title(9791191266054)

'골목의 시간을 그리다'

저자가 두 명 이상일 상황 고려, find_all()메서드로 저자명 담은 \<a>태그 추출: 여러 개 추출되면 for문으로 합쳐 파이썬 리스트로 저장 후 join()메서드로 합치기<br>

발행연도 형식: 2020년 12월 -> re모듈의 **findall()함수**로 원하는 정규식에 매칭되어 문자열 찾아 리스트로 반환할 수 있게 함.<br>
첫 매개변수=원하는 정규식, 두 번째 매개변수=검색 대상 문자열


In [121]:
import re

def get_book_info(row):
    title = row['도서명']
    author = row['저자']
    pub = row['출판사']
    year = row['발행년도']
    # Yes24 도서 검색 페이지 URL
    url = 'http://www.yes24.com/Product/Search?domain=BOOK&query={}'
    # URL에 ISBN을 넣어 HTML 가져오기
    r = requests.get(url.format(row['ISBN']))
    soup = BeautifulSoup(r.text, 'html.parser')   # HTML 파싱
    #뷰티풀수프는 누락된 값만 추출. 정보가 아예 없거나 html요소가 누락된 경우는 오류 발생하여 try~exept문 사용함
    try:
        if pd.isna(title):
            # 클래스 이름이 'gd_name'인 a 태그의 텍스트를 가져오기
            title = soup.find('a', attrs={'class':'gd_name'}) \
                    .get_text()
    except AttributeError:
        pass

    try:
        if pd.isna(author):
            # 클래스 이름이 'info_auth'인 span 태그 아래 a 태그의 텍스트를 가져오기
            authors = soup.find('span', attrs={'class':'info_auth'}) \
                          .find_all('a')
            author_list = [auth.get_text() for auth in authors]
            author = ','.join(author_list)
    except AttributeError:
        pass

    try:
        if pd.isna(pub):
            # 클래스 이름이 'info_auth'인 span 태그 아래 a 태그의 텍스트를 가져오기
            pub = soup.find('span', attrs={'class':'info_pub'}) \
                      .find('a') \
                      .get_text()
    except AttributeError:
        pass

    try:
        if year == -1:
            # 클래스 이름이 'info_date'인 span 태그 아래 텍스트를 가져오기
            year_str = soup.find('span', attrs={'class':'info_date'}) \
                           .get_text()
            # 정규식으로 찾은 값 중에 첫 번째 것만 사용
            year = re.findall(r'\d{4}', year_str)[0]
    except AttributeError:
        pass

    return title, author, pub, year

In [122]:
#누락된 값이 있었던 첫 두행에 방금 만든 함수 적용
#result_type매개변수를 'expand'로 지정하여 반환된 값을 각기 다른 열로 만듦
updated_sample = ns_book5[na_rows].head(2).apply(get_book_info,
    axis=1, result_type ='expand')
updated_sample

0           1           2     3
78    아산 정주영 레거시         김화진  서울대학교출판문화원  2021
265  골목의 시간을 그리다  정명섭.김효찬 지음      초록비책공방  2021

In [123]:
gdown.download('http://bit.ly/3UJZiHw', 'ns_book5_update.csv', quiet=False)

ns_book5_update = pd.read_csv('ns_book5_update.csv', index_col=0)
ns_book5_update.head()

Downloading...
From: http://bit.ly/3UJZiHw
To: /content/ns_book5_update.csv
100%|██████████| 406k/406k [00:00<00:00, 52.3MB/s]


도서명               저자         출판사  발행년도
78                 아산 정주영 레거시              김화진  서울대학교출판문화원  2021
265               골목의 시간을 그리다       정명섭.김효찬 지음      초록비책공방  2021
354                    한국인의 맛           정명섭 지음         추수밭  2021
539              한성부, 달 밝은 밤에           김이삭 지음      고즈넉이엔티  2021
607  100일 완성 마그마 러시아어 중고급 단어장  러포자 구제 연구소 외 지음         문예림  2021

In [124]:
#ns_book5 데이터프레임을 ns_book5_update 데이터프레임 데이터로 업데이트한 후, 누락된 행 개수 확인
ns_book5.update(ns_book5_update)

na_rows = ns_book5['도서명'].isna() | ns_book5['저자'].isna() \
          | ns_book5['출판사'].isna() | ns_book5['발행년도'].eq(-1)
print(na_rows.sum())

4615


In [125]:
#그러고도 누락된 행 제거하여 분석 대상에서 제거
#dropna() 메서드에 '도서명','저자','출판사'열을 리스트로 지정한 후 누락된 값이 있는 행을 삭제
#발행년도 열 값이 -1이 아닌 행만 선택하여 ns_book6 데이터프레임 생성
ns_book5 = ns_book5.astype({'발행년도': 'int32'})

ns_book6 = ns_book5.dropna(subset=['도서명','저자','출판사'])
ns_book6 = ns_book6[ns_book6['발행년도'] != -1]
ns_book6.head()

번호                    도서명                저자    출판사  발행년도           ISBN  \
0   1                인공지능과 흙            김동훈 지음    민음사  2021  9788937444319   
1   2           가짜 행복 권하는 사회            김태형 지음   갈매나무  2021  9791190123969   
2   3  나도 한 문장 잘 쓰면 바랄 게 없겠네            김선영 지음   블랙피쉬  2021  9788968332982   
3   4                예루살렘 해변  이도 게펜 지음, 임재희 옮김  문학세계사  2021  9788970759906   
4   5  김성곤의 중국한시기행 : 장강·황하 편            김성곤 지음    김영사  2021  9788934990833   

  세트 ISBN 부가기호    권 주제분류번호  도서권수  대출건수        등록일자  
0          NaN  NaN    NaN     1     0  2021-03-19  
1          NaN  NaN    NaN     1     0  2021-03-19  
2          NaN  NaN    NaN     1     0  2021-03-19  
3          NaN  NaN    NaN     1     0  2021-03-19  
4          NaN  NaN    NaN     1     0  2021-03-19

In [126]:
#ns_book6 데이터프레임 저장
ns_book6.to_csv('ns_book6.csv', index=False)

In [127]:
#일괄 처리 함수 만들기
def data_fixing(ns_book4):
    """
    잘못된 값을 수정하거나 NaN 값을 채우는 함수

    :param ns_book4: data_cleaning() 함수에서 전처리된 데이터프레임
    """
    # 도서권수와 대출건수를 int32로 바꾸기
    ns_book4 = ns_book4.astype({'도서권수':'int32', '대출건수': 'int32'})
    # NaN인 세트 ISBN을 빈문자열로 바꾸기
    set_isbn_na_rows = ns_book4['세트 ISBN'].isna()
    ns_book4.loc[set_isbn_na_rows, '세트 ISBN'] = ''

    # 발행년도 열에서 연도 네 자리를 추출하여 대체합니다. 나머지 발행년도는 -1로 바꾸기
    ns_book5 = ns_book4.replace({'발행년도':'.*(\d{4}).*'}, r'\1', regex=True)
    unkown_year = ns_book5['발행년도'].str.contains('\D', na=True)
    ns_book5.loc[unkown_year, '발행년도'] = '-1'

    # 발행년도를 int32로 바꾸기
    ns_book5 = ns_book5.astype({'발행년도': 'int32'})
    # 4000년 이상인 경우 2333년을 빼기
    dangun_yy_rows = ns_book5['발행년도'].gt(4000)
    ns_book5.loc[dangun_yy_rows, '발행년도'] = ns_book5.loc[dangun_yy_rows, '발행년도'] - 2333
    # 여전히 4000년 이상인 경우 -1로 바꾸기
    dangun_year = ns_book5['발행년도'].gt(4000)
    ns_book5.loc[dangun_year, '발행년도'] = -1
    # 0~1900년 사이의 발행년도는 -1로 바꾸기
    old_books = ns_book5['발행년도'].gt(0) & ns_book5['발행년도'].lt(1900)
    ns_book5.loc[old_books, '발행년도'] = -1

    # 도서명, 저자, 출판사가 NaN이거나 발행년도가 -1인 행을 찾기
    na_rows = ns_book5['도서명'].isna() | ns_book5['저자'].isna() \
              | ns_book5['출판사'].isna() | ns_book5['발행년도'].eq(-1)
    # 교보문고 도서 상세 페이지에서 누락된 정보를 채우기
    updated_sample = ns_book5[na_rows].apply(get_book_info,
        axis=1, result_type ='expand')
    updated_sample.columns = ['도서명','저자','출판사','발행년도']
    ns_book5.update(updated_sample)

    # 도서명, 저자, 출판사가 NaN이거나 발행년도가 -1인 행을 삭제
    ns_book6 = ns_book5.dropna(subset=['도서명','저자','출판사'])
    ns_book6 = ns_book6[ns_book6['발행년도'] != -1]

    return ns_book6